precompiles/assets-erc20/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let spender: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		{
+			let execution_address = Runtime::AddressMapping::into_account_id(context.address);
+			let asset_id: AssetIdOf<Runtime, Instance> =
+				Runtime::Precompiles::account_to_asset_id(execution_address)
+					.ok_or(error("non-assetId address"))?;
+
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// Dispatch call (if enough gas).
+			// We first cancel any existing approvals
+			// Since we cannot check storage, we need to execute this call without knowing whether
+			// another approval exists already.
+			// But we know that if no approval exists we should get "Unknown"
+			// Allowance() should be checked instead of doing this Result matching
+			let used_gas = match RuntimeHelper::<Runtime>::try_dispatch(
+				<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin::root(),
+				pallet_assets::Call::<Runtime, Instance>::force_cancel_approval(
+					asset_id,
+					Runtime::Lookup::unlookup(caller),
+					Runtime::Lookup::unlookup(spender.clone()),
+				),
+				gasometer.remaining_gas()?,
+			) {
+				Ok(gas_used) => Ok(gas_used),
+				Err(ExitError::Other(e)) => {
+					// This would mean there is not an existing approval
+					// We convert this case to 0 gas used
+					// We could also convert it to one DB read
+					if e.contains("Unknown") {
+						Ok(0)
+					} else {
+						Err(ExitError::Other(e))
+					}
+				}
+				Err(e) => Err(e),
+			}?;
+			gasometer.record_cost(used_gas)?;
+
+			let origin = Runtime::AddressMapping::into_account_id(context.caller);
+
+			// Dispatch call (if enough gas).
+			let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime, Instance>::approve_transfer(
+					asset_id,
+					Runtime::Lookup::unlookup(spender),
+					amount,
+				),
+				gasometer.remaining_gas()?,
+			)?;
+			gasometer.record_cost(used_gas)?;
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_APPROVAL,
+					context.caller,
+					spender,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	fn transfer(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let to: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
```

Comment:
```suggestion
		let amount: BalanceOf<Runtime, Instance> = input.read()?;
```
I believe the utils can convert to the right type for you.